### ISSUE TITLE
[Bug?] Return type in function may contain more object parameters as expected

### ISSUE THREAD

**Issue by dskiba** [microsoft/TypeScript#45534]
# Bug Report

### 🔎 Search Terms

interface/type contain more parameters 

Return type of function may contain more object parameters as expected

When a function return object, passed to function type as generic and that object has more params than required should be an error.  

### 🕗 Version & Regression Information

Tested from 3.6. to current version 4.3.5 

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.3.5#code/C4TwDgpgBAKgYgOwDwGUB8UC8UAUB7AIwCsAuKFASiwxQChaBLBYCAJwDMBDAY2gEkUwTiygBvWlElQWAZ2BkEAVwC2BNrQC+9bngRyocRQm5l4yAUJYZsOOXlYQqmDDnFSoAOi92HAGglSwAAWwjIA8gDWnCBkACwA7P7uMkF4igA2ACYAQhAw4BAAoqys9mQAbACcUAD0NVAEisBQDMAyUAh4wJoU2rr6hKRQFsLQ2GIBkrLyUAmTUClpWbn5kMWlrGQAjLX1nAiZLc0MMgCEmrRAA)

### 💻 Code

```typescript
type TFn<S> = (obj: S) => S

interface IState {
    test: number
}

const Func: TFn<IState> = (store) => ({
    ...store,
    thatsOkay: 47,
    shouldBeTypeError: 69 // but its not
})

const obj: IState =  {
    test: 47
    shouldBeTypeError: 1 // and it is!
}
```

### 🙁 Actual behavior

Expected Func to has an error type since 'TFn' doesn't contain 'shouldBeTypeError' parameter

### 🙂 Expected behavior

Maybe there is a reason for this behavior, but for me, it's not clear why. Object 'obj' has an error, but Func not
Error in 
```typescript
...
shouldBeTypeError: 69
}
...

**Comment by MartinJohns** [comment]
No bug. You didn't specify a return type for your function, so it's inferred. Then assignability is checked, and additional properties to not prevent an assignment, the types are not incompatible. The excess property checks that occurs for object literals is more of a linter feature. This issue comes up here over and over again.

You can get the behaviour you expect by adding a type annotation for the return type to your function: `(store): IState => ({...})`

**Comment by jcalz** [comment]
Object types are open and extendible on purpose; they are not closed or "exact" as proposed in #12936.  Presumably you are expecting excess property checking to happen as in #3823 (a great feature but it leads to widespread confusion about how types work in TypeScript).  But excess property checking only happens on *object literals* being assigned in particular circumstances.  It does not happen with function return values, which means maybe this issue is a duplicate of (checks notes) #241.
